src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -116,5 +138,44 @@ private Uri ConvertFilePathToUri(string filePath)
             var uri = new Uri(filePath, UriKind.Absolute);
             return uri;
         }
+
+        private class SingleThreadedFIFOSemaphoreSlim : IDisposable
+        {
+            private readonly SemaphoreSlim _inner;
+            private readonly ConcurrentQueue<TaskCompletionSource<bool>> _queue;
+
+            public SingleThreadedFIFOSemaphoreSlim()
+            {
+                _inner = new SemaphoreSlim(1, 1);
+                _queue = new ConcurrentQueue<TaskCompletionSource<bool>>();
+            }
+
+            public Task WaitAsync()
+            {
+                var tcs = new TaskCompletionSource<bool>();
+                _queue.Enqueue(tcs);
+
+                _inner.WaitAsync().ContinueWith(_ =>
```

Comment:
Why are we not able to just await this?

Reply:
I wasn't sure what task would be returned to the caller if we awaited this here. This seemed more readable.

Reply:
My question is more geared towards do we need this private class at all? Is there a reason we can't just use a `SemaphoreSlim` (no `TaskCompletionSource` or queue etc.) and `await mySemaphore.WaitAsync()` in the UpdateBuffer method above?

Reply:
"If multiple threads are blocked, there is no guaranteed order, such as FIFO or LIFO, that controls when threads enter the semaphore." - from SemaphoreSlim documentation.

So if multiple threads are waiting at the same time, we want to make sure whichever started blocking first needs to get unblocked first.

Reply:
Ahhh wow. I had no idea semaphoreslim didn't guaruntee that. Really good to know!